src/LinuxCaptureService/ExtractSignalFromMinidump.cpp

```diff
@@ -0,0 +1,103 @@
+
+// Copyright (c) 2022 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "ExtractSignalFromMinidump.h"
+
+#include <string>
+
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/ReadFileToString.h"
+#include "OrbitBase/Result.h"
+
+namespace orbit_linux_capture_service {
+
+namespace {
+
+// The constants and structs below help to interpret the minidump file format. The data here is
+// limited to access the part of the file that we are interested in. For a complete parser see
+// minidump-2-core.cc from https://chromium.googlesource.com/breakpad/breakpad/.
+
+uint32_t kStreamCountOffset = 8;
+uint32_t kStreamDirectoryOffset = 12;
+uint32_t kMdExceptionStream = 6;
+
+struct MDLocationDescriptor {
+  uint32_t data_size;
+  uint32_t rva;
+};
+
+struct MDRawDirectory {
+  uint32_t stream_type;
+  MDLocationDescriptor location;
+};
+
+struct MDException {
+  uint32_t exception_code;
+  uint32_t exception_flags;
+  uint64_t exception_record;
+  uint64_t exception_address;
+  uint32_t number_parameters;
+  uint32_t __align;
+  uint64_t exception_information[15];
+};
+
+struct MDRawExceptionStream {
+  uint32_t thread_id;
+  uint32_t __align;
+  MDException exception_record;
+  MDLocationDescriptor thread_context;
+};
+
+template <typename T>
+ErrorMessageOr<const T*> DataOrError(const std::string& content, uint64_t offset) {
+  if (content.size() < offset + sizeof(T)) {
+    return ErrorMessage("Unexpected end of data.");
+  }
+  return reinterpret_cast<const T*>(content.data() + offset);
+}
+
+template <typename T>
+ErrorMessageOr<const T*> ArrayElementOrError(const std::string& content, uint64_t offset,
+                                             uint64_t index) {
+  return DataOrError<T>(content, offset + index * sizeof(T));
+}
+
+ErrorMessageOr<int32_t> ParseMinidumpForTerminationSignal(const std::string& content) {
+  const int32_t* stream_count = nullptr;
+  OUTCOME_TRY(stream_count, DataOrError<int32_t>(content, kStreamCountOffset));
+  const uint32_t* stream_directory = nullptr;
+  OUTCOME_TRY(stream_directory, DataOrError<uint32_t>(content, kStreamDirectoryOffset));
+  for (int i = 0; i < *stream_count; i++) {
+    const MDRawDirectory* dir_entry = nullptr;
+    OUTCOME_TRY(dir_entry, ArrayElementOrError<MDRawDirectory>(content, *stream_directory, i));
+    if (dir_entry->stream_type != kMdExceptionStream) {
+      continue;
+    }
+    const MDRawExceptionStream* exception_stream = nullptr;
+    OUTCOME_TRY(exception_stream,
+                DataOrError<MDRawExceptionStream>(content, dir_entry->location.rva));
+    return exception_stream->exception_record.exception_code;
+  }
+  return -1;
+}
+
+}  // namespace
+
+int32_t ExtractSignalFromMinidump(const std::filesystem::path& path) {
```

Comment:
Since we're already using "ErrorMessageOr" in the file, maybe we should also return it here? This could reduce the code below to two  "OUTCOME_TRY()" lines and the tests could test for the exact failure.

Reply:
In case `ErrorMessageOr` is too much (e.g., we don't need to pass a string outside), I would prefer `std::optional<int32_t>` (or maybe `std::optional<uint32_t>`, although I think signals are `int`s).

Reply:
I went with `ErrorMessageOr`